File: frontend/src/utils/validators.js
export function validateSnils(value) {
  const digits = (value || '').replace(/\D/g, '');
  if (digits.length !== 11) return 'СНИЛС должен состоять из 11 цифр.';

  const number = parseInt(digits, 10);
  if (number <= 1001998) return null;

  let sum = 0;
  for (let i = 0; i < 9; i++) sum += Number(digits[i]) * (9 - i);

  let checkSum;
  if (sum < 100) checkSum = sum;
  else if (sum === 100 || sum === 101) checkSum = 0;
  else {
    checkSum = sum % 101;
    if (checkSum === 100 || checkSum === 101) checkSum = 0;
  }

  return checkSum !== Number(digits.slice(9)) ? 'Неверное контрольное число СНИЛС.' : null;
}

export function validateFutureDate(value) {
  if (!value) return 'Укажите желаемую дату приёма.';
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  if (new Date(value) < today) return 'Дата не может быть в прошлом.';
  return null;
}

const FIO_RE = /^[А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)?(\s[А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)?){1,2}$/;

export function validateFio(value) {
  if (!FIO_RE.test((value || '').trim())) {
    return 'ФИО должно быть на русском языке: фамилия и имя (отчество не обязательно).';
  }
  return null;
}

export function validatePhone(value) {
  const digits = (value || '').replace(/\D/g, '');
  if (digits.length !== 11 || !digits.startsWith('7')) {
    return 'Укажите телефон в формате +7 (XXX) XXX-XX-XX.';
  }
  return null;
}

File: backend/journal/validators.py
import re
from django.core.exceptions import ValidationError
from django.utils import timezone


def validate_snils(value):
    digits = re.sub(r'\D', '', value)

    if len(digits) != 11:
        raise ValidationError("СНИЛС должен состоять из 11 цифр.")

    if int(digits) <= 1001998:
        return

    sum_val = sum(int(digits[i]) * (9 - i) for i in range(9))
    check_sum = int(digits[9:])

    if sum_val < 100:
        calc_sum = sum_val
    elif sum_val in (100, 101):
        calc_sum = 0
    else:
        calc_sum = sum_val % 101
        if calc_sum in (100, 101):
            calc_sum = 0

    if calc_sum != check_sum:
        raise ValidationError("Неверное контрольное число СНИЛС.")


def validate_future_date(value):
    if value < timezone.now().date():
        raise ValidationError("Желаемая дата приёма не может быть в прошлом.")


FIO_RE = re.compile(r'^[А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)?(\s[А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)?){1,2}$')


def validate_fio(value):
    if not FIO_RE.match(value.strip()):
        raise ValidationError("ФИО должно быть на русском языке: фамилия и имя (отчество не обязательно).")


def validate_phone(value):
    digits = re.sub(r'\D', '', value)
    if len(digits) != 11 or not digits.startswith('7'):
        raise ValidationError("Телефон должен быть в формате +7 (XXX) XXX-XX-XX.")